notion-core/src/main/scala/com/ambiata/notion/core/SequenceFileIO.scala

```diff
@@ -0,0 +1,118 @@
+
+  def readLocation(location: Location, client: AmazonS3Client): RIO[List[A]] = for {
+    in <- (location match {
+      case HdfsLocation(p) =>
+        RIO.safe[InputStream](FileSystem.get(conf).open(new Path(p)))
+      case LocalLocation(p) =>
+        RIO.safe[InputStream](new LocalInputStream(p))
+      case S3Location(bucket, key) =>
+        S3InputStream.stream(S3Address(bucket, key), client)
+    })
+    as <- readInputStream(in)
+  } yield as
+
+  def readInputStream(input: InputStream): RIO[List[A]] =
+    RIO.using(RIO.safe(new FSDataInputStream(input))) {
+      in => {
+        RIO.using(RIO.safe(new SequenceFile.Reader(conf, SequenceFile.Reader.stream(in)))) {
+          reader => for {
+            _ <- compatibleClassOrFail(reader.getKeyClass, km.runtimeClass, "key")
+            _ <- compatibleClassOrFail(reader.getValueClass, vm.runtimeClass, "value")
+            r <- RIO.safe[List[A]] {
+              val buffer = scala.collection.mutable.ListBuffer[A]()
+              while (SequenceUtil.nextSafe(reader, emptyKey, emptyValue)) {
+                val a: A  = fromKeyValue(emptyKey, emptyValue)
+                buffer += a
+              }
+              buffer.toList
+            }
+          } yield r
+        }
+      }
+    }
+
+  def writeLocation(location: Location, client: AmazonS3Client, codec: Option[CompressionCodec], as: List[A]): RIO[Unit] = for {
+    out <- (location match {
+      case HdfsLocation(p) =>
+        Hdfs.mkdir(new Path(p).getParent).run(conf) >>
+          RIO.safe[OutputStream](FileSystem.get(conf).create(new Path(p)))
+      case LocalLocation(p) =>
+        Directories.mkdirs(FilePath.unsafe(p).dirname) >>
+          RIO.safe[OutputStream](new BufferedOutputStream(new FileOutputStream(p)))
+      case S3Location(bucket, key) =>
+        S3OutputStream.stream(S3Address(bucket, key), client)
+    })
+    _ <- writeOutputStream(out, codec, as)
+  } yield ()
+
+  def writeOutputStream(output: OutputStream, codec: Option[CompressionCodec], as: List[A]): RIO[Unit] = for {
+    _    <- RIO.using(RIO.safe(new FSDataOutputStream(output, new FileSystem.Statistics("SequenceFile")))) {
+      out => {
+        val opts = List(
+          SequenceFile.Writer.stream(out),
+          SequenceFile.Writer.keyClass(km.runtimeClass),
+          SequenceFile.Writer.valueClass(vm.runtimeClass)
+        ) ++ codec.map {
+          SequenceFile.Writer.compression(SequenceFile.CompressionType.BLOCK, _)
+        }
+        RIO.using(RIO.safe(SequenceFile.createWriter(conf, opts: _*))) {
+          writer => RIO.safe(as.map(toKeyValue).foreach((writer.append _).tupled))
+        }
+      }
+    }
+  } yield ()
+
+  def compatibleClassOrFail(fileClass: Class[_], expectedClass: Class[_], label: String): RIO[Unit] =
+    if(!expectedClass.isAssignableFrom(fileClass))
+      RIO.fail(s"Expecting ${label} class in SequenceFile to be '${expectedClass.getSimpleName}', but found '${fileClass.getSimpleName}'")
```

Comment:
We need to start getting better with these errors, should not be RIO fail. Top level methods should have type of `RIO[ReadError \/ A]` and then there can be a `read/write{OrFail}` combinators if you really want to do this, but it shouldn't be the default.